Don't return TEXT type for functions that take TEXT

craigtaverner opened the pull request:
Always return `KEYWORD` for functions that previously returned `TEXT`, because any change to the value, no matter how small, is enough to render meaningless the original analyzer associated with the `TEXT` field value. In principle, if the attribute is no longer the original `FieldAttribute`, it can no longer claim to have the type `TEXT`.

This has been done for all functions: conversion functions, aggregating functions, multi-value functions. There were several that already produced `KEYWORD` for `TEXT` input (eg. ToString, FromBase64 and ToBase64, MvZip, ToLower, ToUpper, DateFormat, Concat, Left, Repeat, Replace, Right, Split, Substring), but many others that incorrectly claimed to produce `TEXT`, while this was really a false claim. This PR makes that now strict, and includes changes to the functions' units tests to disallow the tests to expect any functions output to be `TEXT`.

One side effect of this change is that methods that take multiple parameters that require all of them to have the same type, will now treat TEXT and KEYWORD the same. This was already the case for functions like `Concat`, but is now also the case for `Greatest`, `Least`, `Case`, `Coalesce` and `MvAppend`.

An associated change is that the type casting operator `::text` has been entirely removed. It used to map onto the `ToString` function which returned type KEYWORD, and so `::text` really produced a `KEYWORD`, which is a lie, or at least a `bug`, which is now fixed. Should we ever wish to actually produce real `TEXT`, we might love the fact that this operator has been freed up for future use (although it seems likely that function will require parameters to specify the analyzer, so might never be an operator again).

### Backwards compatibility issues:

This is a change that will fail BWC tests, since we have many tests that assert on TEXT output to functions. For this reason we needed to block two scenarios:

* We used the capability `functions_never_emit_text` to prevent 7 csv-spec tests and 2 yaml tests from being run against older versions that still emit text.
* We used `skipTest` to also block those two yaml tests from being run against the latest build, but using older yaml files downloaded (as far back as 8.14).

In all cases the change observed in these tests was simply the results columns no longer having `text` type, and instead being `keyword`.

### Kibana

Kibana makes use of some JSON files we produce during tests runs that explain the signatures of the functions, and the available type-casting operators. These are now edited to remove all `TEXT` output, and remove the `::test` operator.

### Fixes

Fixes #114333
Fixes #111537

craigtaverner committed:
Don't return TEXT type for functions that take TEXT

Always return KEYWORD, because any change to the value, no matter how small, is enough to render meaningless the original analyzer associated with the TEXT value.

elasticsearchmachine commented:
Pinging @elastic/es-analytical-engine (Team:Analytics)

craigtaverner committed:
Update docs/changelog/114334.yaml

craigtaverner committed:
Remove text from return type documented, and meta-tests

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into esql_functions_never_return_text

craigtaverner committed:
Support REVERSE

craigtaverner committed:
Merge branch 'esql_functions_never_return_text' of github.com:craigtaverner/elasticsearch into esql_functions_never_return_text

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into esql_functions_never_return_text

craigtaverner committed:
Added EsqlCapabilities for reverse function emitting keyword

craigtaverner committed:
Fixed failing test

astefan reviewed (changes requested):
I think it's a good step forward, but imo we could do better.
This needs a general evaluation of the situation because `text` as an output data type of our own functions is confusing, meaningless and not so useful at this point in the life of the language. So:
- all functions need to be checked. I've caught `max(text)` which returns `text`. If we are to make this change, let's make it general.
- to catch any misses, we should have a check in a central location for return data types of any kind of function/operator/anything-else (maybe after the logical planner is done, not sure). Maybe with an `assert` so that we don't fail queries at runtime, unless it's CI/tests.
- another way of looking at the problem is that `text` (as a data type) does only make sense for FieldAttributes? (meaning we can accept `text` fields, we can do stuff with them, but creating a `text` shouldn't be possible)

astefan commented:
> > * Change the parent class `UnaryScalarFunction` to always return KEYWORD for TEXT fields (done in this PR to cover functions not yet fixed)
> 
> I think I'd _prefer_ if `UnaryScalarFunction` didn't have a default return type. That's a bigger change though because everyone's got to fill in their return type.

++
Also, `UnaryScalarFunction` is not the only function in the class' hierarchy.

nik9000 commented:
> central location

I was thinking that the tests for functions could assert that they never spit out `text`. I think you could override that when we go to implement `ANALYZE(kwd, FRENCH)` or whatever we call it, but only a hand full of function will ever output `text` fields and I'm ok with extra work telling you "you probably don't want to return `text` here" when writing them.

luigidellaquila committed:
Fix tests

craigtaverner committed:
Expand EsqlCapability to cover all functions now emitting KEYWORD

astefan commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/aggregate/Values.java:
```diff
@@ -80,7 +82,8 @@ public Values replaceChildren(List<Expression> newChildren) {
     @Override
     public DataType dataType() {
-        return field().dataType();
+        DataType t = field().dataType();
```
Instead of doing this for every aggregation function, how about going to `AggregateFunction` and define there something similar to `UnaryScalarFunction.dataType()`?

craigtaverner committed:
More tests failing due to new behaviour

craigtaverner committed:
Remove type assertion for function merged back to 8.x already

We pull this text from 8.x and run it on current builds in yamlRestCompatTest, so we need to disable this check here (and in 8.x)

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into esql_functions_never_return_text

craigtaverner committed:
Mute yamlRestCompatTest for tests that assert on TEXT/KEYWORD columns

craigtaverner committed:
Re-enable KEYWORD assertion on reverse text, but not values

Since values is available for many versions, disabling the test for all but the latest version seems counter-productive, so I rather just remove the type assertions.

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into esql_functions_never_return_text

craigtaverner committed:
Cannot skip tests named "values"

This seems to trigger a classcast exception inside Jackson!

class com.fasterxml.jackson.databind.node.IntNode cannot be cast to class com.fasterxml.jackson.databind.node.ArrayNode (com.fasterxml.jackson.databind.node.IntNode and com.fasterxml.jackson.databind.node.ArrayNode are in unnamed module of loader org.gradle.internal.classloader.VisitableURLClassLoader$InstrumentingVisitableURLClassLoader @50337c96)

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into esql_functions_never_return_text

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into esql_functions_never_return_text

ioanatia commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/aggregate/Max.java:
```diff
@@ -107,7 +109,8 @@ protected TypeResolution resolveType() {
     public DataType dataType() {
-        return field().dataType();
+        DataType t = field().dataType();
+        return t == TEXT ? KEYWORD : t;
```
I have a small suggestion - feel free to leave this as it is - instead of checking for `t == TEXT` we could do `DataType.isString(t) ? KEYWORD : t`. The intention would be a bit clearer that for string arguments, the return type is keyword.
This would work in case we want to add another string type - which we do, we are working on adding `semantic_text`. So I think that even if we leave this as it is for now, we would still need to change this check to use `DataType.isString` when we add support for `semantic_text` to be a string type.

craigtaverner replied:
I double-checked, and this particular case was indeed not switched to use `noText()`, so I'll fix that directly, and look for more cases I might have missed.

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into esql_functions_never_return_text

craigtaverner committed:
Remove support for ::text type conversion

This conversion lied, and always produced KEYWORD. Since we're removing all support for producing fake TEXT everywhere, it is only logical that this 'bug' get fixed too.

craigtaverner committed:
Simplify the code to check if a return type is Text or not

craigtaverner committed:
Fix failing test

craigtaverner committed:
Remove TEXT output from CASE, GREATEST, LEAST

In these cases, the incoming TEXT is interpreted as KEYWORD.
This has a particular effect on CASE because we now support mixing KEYWORD and TEXT in incoming arguments, and all combinations return KEYWORD.

This PR also sets all expectedType values in all TestCase objects to KEYWORD if they were TEXT, so this should block any functions tests from allowing TEXT output.

craigtaverner committed:
Remove TEXT output from COALESCE

craigtaverner committed:
Remove TEXT output from MV_APPEND

craigtaverner committed:
Remove TEXT output from MV_SLICE and MV_SORT

craigtaverner committed:
Fixed failing tests by basing new TEX/KEYWORD combinations on a simpler non-folding case

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into esql_functions_never_return_text

craigtaverner committed:
Fixed remaining ternary check for TEXT

craigtaverner committed:
Added fixed json after running all esql tests locally

elasticsearchmachine commented:
Pinging @elastic/kibana-esql (ES|QL-ui)

astefan reviewed (approved):
LGTM. Looking much better. Thanks, Craig.

luigidellaquila reviewed (approved):
LGTM, thanks!

craigtaverner commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/type/EsqlDataTypeConverter.java:
```diff
@@ -117,7 +117,6 @@ public class EsqlDataTypeConverter {
         entry(LONG, ToLong::new),
         // ToRadians, typeless
         entry(KEYWORD, ToString::new),
-        entry(TEXT, ToString::new),
```
Yes! I was busy creating a new issue for that, but glad to see you already did, way back! I'll add a 'fixes' line to the PR description.

craigtaverner committed:
Update docs/changelog/114334.yaml

elasticsearchmachine commented:
Hi @craigtaverner, I've updated the changelog YAML for you.

craigtaverner commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/TestCaseSupplier.java:
```diff
@@ -1435,7 +1435,7 @@ public static TestCase typeError(List<TypedData> data, String expectedTypeError)
             this.data = data;
             this.evaluatorToString = evaluatorToString;
-            this.expectedType = expectedType;
+            this.expectedType = expectedType == null ? null : expectedType.noText();
```
This is the comprehensive 'catch all' case we added late in the PR once we had covered all the low hanging fruit. But perhaps this means we can remove some of the other earlier code. I'll investigate.

craigtaverner replied:
Indeed there were 12 places we had added this `noText()` that no longer needed it, since they all called into this TestCase constructor. I've removed them all and will push a commit after re-running tests to be safe.

craigtaverner committed:
Removed unnecessary calls to noText in test cases

Since we added a check for this inside the TestCase constructor, all these extra calls are no longer needed.

craigtaverner committed:
Merge branch 'esql_functions_never_return_text' of github.com:craigtaverner/elasticsearch into esql_functions_never_return_text

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into esql_functions_never_return_text

bpintea commented on x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/type/DataType.java:
```diff
@@ -584,6 +584,10 @@ static Builder builder() {
         return new Builder();
     }
 
+    public DataType noText() {
```
`textAsKeyword()`? `noText` is compact, but I find it a tad misleading. Minor preference, can stay as is.

craigtaverner replied:
The idea is sound, but I have a minor preference for `noText` as `textAsKeyword` also seems to say something about the current datatype being TEXT; which it likely is not.

bpintea commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/aggregate/Top.java:
```diff
@@ -175,7 +177,8 @@ protected TypeResolution resolveType() {
     @Override
     public DataType dataType() {
-        return field().dataType();
+        DataType t = field().dataType();
+        return t == TEXT ? KEYWORD : t;
```
noText()?

craigtaverner replied:
done

bpintea commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/aggregate/Values.java:
```diff
@@ -91,7 +93,8 @@ public Values withFilter(Expression filter) {
     public DataType dataType() {
-        return field().dataType();
+        DataType t = field().dataType();
+        return t == TEXT ? KEYWORD : t;
```
noText()?

craigtaverner replied:
done

bpintea reviewed (approved):
Lgtm

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into esql_functions_never_return_text

craigtaverner committed:
Removed three remaining ternary functions, using noText instead

craigtaverner commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/aggregate/Min.java:
```diff
@@ -107,7 +109,8 @@ protected TypeResolution resolveType() {
     @Override
     public DataType dataType() {
-        return field().dataType();
+        DataType t = field().dataType();
+        return t == TEXT ? KEYWORD : t;
```
Done. I normally like to accept the commit suggestion, but in this case the remaining imports would cause a build failure, so I did this locally and will push a commit for all three places soon.

bpintea replied:
all good, it's not like i really contributed :)

craigtaverner commented on x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/type/DataType.java:
```diff
@@ -584,6 +584,10 @@ static Builder builder() {
     }
 
+    public DataType noText() {
+        return this == DataType.TEXT ? DataType.KEYWORD : this;
```
Done

elasticsearchmachine commented:
## 💚 Backport successful
| Status | Branch | Result |
|:------:|:------:|:------:|
| ✅ |  [8.x](https://github.com/elastic/elasticsearch/pull/115625)  | [<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/115625">](https://github.com/elastic/elasticsearch/pull/115625) |
